File: src/main/java/march_2025/NumDistinct.java
package march_2025;

public class NumDistinct {

    public static void main(String[] args) {
        String s = "rabbbit";
        String t = "rabbit";

        NumDistinct numDistinct = new NumDistinct();
        int result = numDistinct.numDistinct(s, t);
        System.out.println(result);
    }

    int[][] dp;
    public int numDistinct(String s, String t) {
        if (s.length() < t.length()) {
            return 0;
        }
        dp = new int[s.length() + 1][t.length() + 1];
        for (int i = 0; i < dp.length; i++) {
            for (int j = 0; j < dp[i].length; j++) {
                dp[i][j] = -1;
            }
        }

        return dfs(s, t, 0, 0);
    }

    private int dfs(String s, String t, int i, int j) {
        if (j == t.length()) {
            return 1;
        }
        if (i == s.length()) {
            return 0;
        }
        if (dp[i][j] != -1) {
            return dp[i][j];
        }
        if (s.charAt(i) == t.charAt(j)) {
            dp[i][j] = dfs(s, t, i + 1, j + 1) + dfs(s, t, i + 1, j);
        } else {
            dp[i][j] = dfs(s, t, i + 1, j);
        }
        return dp[i][j];
    }

}

File: python/num_distinct.py
class Solution(object):
    # O(n * m) time | O(n * m) space
    def numDistinct(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: int
        """
        if len(s) < len(t):
            return 0
        
        dp = [ [-1] * (len(s) + 1) for _ in range(len(t) + 1) ]

        def dfs(i, j):
            if j == len(t):
                return 1
            
            if i == len(s):
                return 0
            
            if dp[i][j] != -1:
                return dp[i][j]
            
            res = 0
            if s[i] == t[j]:
                res += dfs(i + 1, j + 1)
                res += dfs(i + 1, j)
            else:
                res += dfs(i + 1, j)
            dp[i][j] = res
            return res

        return dfs(0, 0)
    
solution = Solution()
s = "rabbbit"
t = "rabbit"
print(solution.numDistinct(s, t))